spock-core/src/main/java/org/spockframework/runtime/ParameterizedSpecRunner.java

```diff
@@ -132,20 +136,45 @@ private int estimateNumIterations(Object[] dataProviders) {
     return result == Integer.MAX_VALUE ? -1 : result;
   }
 
-  private void runIterations(Iterator[] iterators, int estimatedNumIterations) {
-    if (runStatus != OK) return;
+  private void runIterations(final FeatureInfo currentFeature, final Iterator[] iterators, final int estimatedNumIterations) {
+    if (runStatus.get() != OK) return;
+
+    boolean atLeastOneIteration = false;
+
+    Scheduler schedulerForRunIterations = this.scheduler.deriveScheduler(!currentFeature.isReportIterations() || !currentFeature.isSupportParallelExecution());
+
+    while (haveNext(currentFeature, iterators)) {
+      atLeastOneIteration = true;
+      final Object[] dataValues = nextArgs(currentFeature, iterators);
 
-    while (haveNext(iterators)) {
-      initializeAndRunIteration(nextArgs(iterators), estimatedNumIterations);
+      if (haveNext(currentFeature, iterators)) {
+        schedulerForRunIterations.schedule(new Runnable() {
+          @Override
+          public void run() {
+            runStatus.set(OK);
```

Comment:
@leonard84 I can not understand you. 
1) runStatus is local for Specification
2) features are independent (same for iterations) - they will not (and should not) affect each other. Statuses of other features does not affect status of current.
3) yes, runStatus is bad decision. I do not like it too. But it was introduced before my changes and I do not want to remove it in this PR.